### DISCUSSION TITLE
How to Refine mesh in Moose

### DISCUSSION THREAD

**Discussion by Ali1990dashti** [idaholab/moose#18655] [Q&A Meshing] [answered]
Dear community,
I have a mesh file which has some volumes and their related physical groups. I want to refine my mesh in one of the physical groups. I have seen that in moose we can define a box and then refine mesh in that box. But, is it possible to refine the mesh in the volume using its name or block number rather than a box with coordinates? I tried the following code to refine in the box but it was not successful for me (I want to for example split each tetrahedron to three ones in one of my physical groups (the Reservoir group)):
```
[Adaptivity]
 max_h_level = 3
 cycles_per_step = 1
 marker = box
 [./Markers]
   [./box]
     type = BoxMarker
     bottom_left = '1800. 1200. -1950.'
     top_right = '2500. 800. -1500.'
     inside = REFINE
     outside =  DO_NOTHING
   [../]
 [../]
[]
```
I have seen explanations in [this page](https://mooseframework.inl.gov/syntax/Adaptivity/index.html) but still I could not solve my issue. I have also uploaded my mesh file.
I do appreciate any consideration in advance.
[mesh.zip](https://github.com/idaholab/moose/files/7021608/mesh.zip)

**Comment by aeslaughter** [comment, marked as answer]
The UniformMarker, as the name suggests, can perform a uniform refinement within a subdomain by setting the "block" parameter.
https://mooseframework.inl.gov/source/markers/UniformMarker.html

**Reply by aeslaughter** [reply]
Actually, all markers have the "block" parameter so they should be able to be restricted to a domain.

**Reply by Ali1990dashti** [reply]
Dear @aeslaughter, Thanks for devoting time to my issue. I tried the following method you proposed but it did not converge. My problem is a steady state one and I just want to refine mesh in one block and then solve the steady state problem. It firstly converges in solving the problem and then goes for adaptavity but in this step it does not converge. it says : `Linear solve did not converge due to DIVERGED_BREAKDOWN iterations 30`.

**Reply by GiudGiud** [reply]
Can you cut the simulation there and look at the mesh? Did the refinement work as expected?

**Reply by Ali1990dashti** [reply]
When I use the `--mesh-only` functionality, it gives me the mesh but it is not refined. I do not know why it firstly runs the simulation and then starts to adapt the mesh. I expected that firstly it should refine the mesh and then do all the numerical calculations.

**Reply by GiudGiud** [reply]
Adaptivity does not run with --mesh-only. You need to look at the output of the mesh  after the simulation. 
Since the simulation fails after adaptivity:
- try to relax the tolerances enough so that you still get a 'simulation converged' and regular output
OR
- change the output execute_on flags, that you still get the output right after adaptivity

**Reply by aeslaughter** [reply]
I am not able to get the adaptivity working like I would expect. I am creating an issue with an input file and will see if anyone knows what is missing or if there is a problem.

**Reply by Ali1990dashti** [reply]
I exported the incomplete simulation using flags in the output block of my input file. In paraview I can see the selected block (I called it `const` ), but its mesh size has not changed. I set my parameters as following:
```
[Adaptivity]
  [./Markers]
    [./const]
      type =  UniformMarker
      mark = REFINE
      block = 'Reservoir'
    [../]
  [../]
  marker = const
  steps = 2
[]
```
It also creates a file with this prefix which is new for me: `.e-s002`. But still it does not converge.

**Reply by GiudGiud** [reply]
Sounds good. 
Otherwise, @socratesgorilla is working on a block refinement using mesh generators. This will be much easier to use than adaptivity in my opinion. It's coming real soon, PR next week probably

**Reply by aeslaughter** [reply]
I figured out the correct syntax, see #18657.

**Reply by aeslaughter** [reply]
> It also creates a file with this prefix which is new for me: `.e-s002`. But still it does not converge.

Every time the mesh changes a new Exodus file is created.
